go/cmd/vpnkit-forwarder/many.go

```diff
@@ -46,72 +45,13 @@ func manyPorts() {
 					log.Printf("Error accepting connection: %#v", err)
 					return // no more listening
 				}
-				go func(conn net.Conn) {
-					// Read header which describes TCP/UDP and destination IP:port
-					d, err := unmarshalDestination(conn)
-					if err != nil {
-						log.Printf("Failed to unmarshal header: %#v", err)
-						conn.Close()
-						return
-					}
-					switch d.Proto {
-					case TCP:
-						backendAddr := net.TCPAddr{IP: d.IP, Port: int(d.Port), Zone: ""}
-						libproxy.HandleTCPConnection(conn.(libproxy.Conn), &backendAddr, quit)
-						break
-					case UDP:
-						backendAddr := &net.UDPAddr{IP: d.IP, Port: int(d.Port), Zone: ""}
-
-						proxy, err := libproxy.NewUDPProxy(backendAddr, libproxy.NewUDPConn(conn), backendAddr)
-						if err != nil {
-							log.Printf("Failed to setup UDP proxy for %s: %#v", backendAddr, err)
-							conn.Close()
-							return
-						}
-						proxy.Run()
-						break
-					default:
-						log.Printf("Unknown protocol: %d", d.Proto)
-						conn.Close()
-						return
-					}
-				}(conn)
+				defer conn.Close()
```

Comment:
This isn't right though, cause now we aren't closing in non-error case, to fix that we could add another goroutine with an explicit return, but I'm not too sure what is the intention here.

To be honest, it seems like this would become easier if we get rid of the `forever` channel and have single listener variable instead of a slice, what do you think?

Reply:
I agree -- since there is only one working listener (vsock or hvsock but not both) we should simplify.